storage/series/reader.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package series
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3db/context"
+	"github.com/m3db/m3db/retention"
+	"github.com/m3db/m3db/storage/block"
+	"github.com/m3db/m3db/ts"
+	xio "github.com/m3db/m3db/x/io"
+	xerrors "github.com/m3db/m3x/errors"
+)
+
+var (
+	errSeriesReadInvalidRange = errors.New(
+		"series invalid time range read argument specified")
+)
+
+// Reader reads results from a series, or a series block
+// retriever or both.
+// It is implemented as a struct so it can be allocated on
+// the stack.
+type Reader struct {
+	opts        Options
+	cloneableID ts.ID
+	retriever   QueryableBlockRetriever
+}
+
+// NewReaderForRetriever returns a reader for a series
```

Comment:
I'm a little confused why we need this new reader. Aren't we currently serving reads from these block retrievers already? Why do we need something new to support peer bootstrapping?

Reply:
So this was because we don't want to create a `Series` when we want to service a read for a series that isn't in memory.  It's moving the stitching of results together from the `Series` to an independent party that can be used directly in the shard (without needing to create a `Series` when one doesn't exist in memory).

Reply:
Ok cool, possibly worth a comment, but maybe not. Kind of requires a wholistic understanding of whats going on